File: tools/fetch_buildings.js
const fs = require('fs');

// 備份機制
fs.copyFileSync('campus_nodes_edges.json', 'campus_nodes_edges_backup.json');
console.log("✅ 已備份原始資料");

const rawData = fs.readFileSync('campus_nodes_edges.json');
const campusData = JSON.parse(rawData);

// 1. 測量地球兩點真實距離 (公尺)
function getDistance(lat1, lon1, lat2, lon2) {
  const R = 6371000;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a = Math.sin(dLat/2) * Math.sin(dLat/2) +
            Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) *
            Math.sin(dLon/2) * Math.sin(dLon/2);
  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1-a));
  return R * c;
}

// 2. 向量投影數學：尋找點 P 在線段 AB 上的「垂直投影點」
function getProjectedPoint(pLat, pLng, aLat, aLng, bLat, bLng) {
  const dx = bLng - aLng;
  const dy = bLat - aLat;
  const lengthSquared = dx * dx + dy * dy;

  if (lengthSquared === 0) return { lat: aLat, lng: aLng, t: 0 };

  // 計算 t 值 (投影點在線段上的比例)
  let t = ((pLng - aLng) * dx + (pLat - aLat) * dy) / lengthSquared;
  t = Math.max(0, Math.min(1, t)); // 限制在 0~1 之間，確保不超出線段兩端

  return {
    lat: aLat + t * dy,
    lng: aLng + t * dx,
    t: t
  };
}

async function fetchAndConnectBuildings() {
  console.log("⏳ 啟動 Snap-to-Road 垂直吸附抓取...");
  const bbox = "23.550,120.460,23.570,120.480";
  const query = `
    [out:json][timeout:25];
    (
      way["building"](${bbox});
      relation["building"](${bbox});
      node["amenity"="restaurant"](${bbox});
      node["amenity"="cafe"](${bbox});
      node["shop"="convenience"](${bbox});
    );
    out center;
  `;

  const url = "https://overpass.kumi.systems/api/interpreter";

  try {
    const response = await fetch(url, {
      method: "POST",
      body: "data=" + encodeURIComponent(query),
      headers: { "Content-Type": "application/x-www-form-urlencoded" }
    });

    const osmData = await response.json();
    
    // 建立 Node 快速查詢字典
    const nodeMap = new Map();
    campusData.nodes.forEach(n => nodeMap.set(n.id, n));

    let splitEdgesCount = 0;
    let nextVirtualNodeId = 9000001; // 從 900 萬開始編號，代表我們切出來的虛擬點

    osmData.elements.forEach(element => {
      const name = element.tags && element.tags.name ? element.tags.name : null;
      if (!name) return;

      const lat = element.lat || (element.center && element.center.lat);
      const lng = element.lon || (element.center && element.center.lon);
      if (!lat || !lng) return;

      const newId = element.id;

      // 新增店家/建築物 Node
      const poiNode = {
        id: newId,
        name: name,
        lat: Number(lat),
        lng: Number(lng),
        accessible: true,
        type: "poi"
      };
      campusData.nodes.push(poiNode);
      nodeMap.set(newId, poiNode);

      // 🎯 核心邏輯：這次我們不找最近的 Node，而是找最近的「Edge (線段)」！
      let minDistance = Infinity;
      let bestEdgeInfo = null;
      let bestProjectedPoint = null;

      campusData.edges.forEach(edge => {
        if (edge._deleted) return; // 跳過已經被剪斷的廢棄路線
        
        const nodeA = nodeMap.get(edge.from);
        const nodeB = nodeMap.get(edge.to);
        
        if (!nodeA || !nodeB) return;
        if (nodeA.type === 'poi' || nodeB.type === 'poi') return; // 不要連到別家店身上

        // 算出垂直點
        const proj = getProjectedPoint(lat, lng, nodeA.lat, nodeA.lng, nodeB.lat, nodeB.lng);
        // 算出店家到垂直點的距離
        const dist = getDistance(lat, lng, proj.lat, proj.lng);

        if (dist < minDistance) {
          minDistance = dist;
          bestEdgeInfo = { edge, nodeA, nodeB };
          bestProjectedPoint = proj;
        }
      });

      // 如果找到的最近道路在 150 公尺內，執行「吸附」
      if (bestEdgeInfo && minDistance < 150) {
        let connectionNodeId;

        // 狀況 1：投影點剛好在端點上
        if (bestProjectedPoint.t === 0) {
          connectionNodeId = bestEdgeInfo.nodeA.id;
        } else if (bestProjectedPoint.t === 1) {
          connectionNodeId = bestEdgeInfo.nodeB.id;
        } 
        // 狀況 2：投影點在線段中間！拿出剪刀！
        else {
          connectionNodeId = nextVirtualNodeId++;
          
          // 生出一個虛擬路口 Node
          const virtualNode = {
            id: connectionNodeId,
            name: "", // 路口沒有名字
            lat: Number(bestProjectedPoint.lat.toFixed(6)),
            lng: Number(bestProjectedPoint.lng.toFixed(6)),
            accessible: bestEdgeInfo.edge.accessible,
            type: "virtual_path"
          };
          campusData.nodes.push(virtualNode);
          nodeMap.set(connectionNodeId, virtualNode);

          // 計算剪斷後的兩段新長度
          const distA_to_V = getDistance(bestEdgeInfo.nodeA.lat, bestEdgeInfo.nodeA.lng, virtualNode.lat, virtualNode.lng);
          const distV_to_B = getDistance(virtualNode.lat, virtualNode.lng, bestEdgeInfo.nodeB.lat, bestEdgeInfo.nodeB.lng);

          // 標記原本的線段為「已刪除」(包含正向跟反向)
          bestEdgeInfo.edge._deleted = true;
          const reverseEdge = campusData.edges.find(e => e.from === bestEdgeInfo.nodeB.id && e.to === bestEdgeInfo.nodeA.id && !e._deleted);
          if (reverseEdge) reverseEdge._deleted = true;

          // 重新接線：A <-> 虛擬點 <-> B
          campusData.edges.push(
            { from: bestEdgeInfo.nodeA.id, to: virtualNode.id, distance: Number(distA_to_V.toFixed(8)), accessible: bestEdgeInfo.edge.accessible },
            { from: virtualNode.id, to: bestEdgeInfo.nodeA.id, distance: Number(distA_to_V.toFixed(8)), accessible: bestEdgeInfo.edge.accessible },
            { from: virtualNode.id, to: bestEdgeInfo.nodeB.id, distance: Number(distV_to_B.toFixed(8)), accessible: bestEdgeInfo.edge.accessible },
            { from: bestEdgeInfo.nodeB.id, to: virtualNode.id, distance: Number(distV_to_B.toFixed(8)), accessible: bestEdgeInfo.edge.accessible }
          );
          splitEdgesCount++;
        }

        // 把店家連到我們決定好的 connectionNodeId 上 (雙向)
        campusData.edges.push(
          { from: newId, to: connectionNodeId, distance: Number(minDistance.toFixed(8)), accessible: true },
          { from: connectionNodeId, to: newId, distance: Number(minDistance.toFixed(8)), accessible: true }
        );
      }
    });

    // 最後，把那些被我們剪斷標記為 _deleted 的舊線段徹底丟掉
    campusData.edges = campusData.edges.filter(e => !e._deleted);

    fs.writeFileSync('campus_nodes_edges.json', JSON.stringify(campusData, null, 2));
    
    console.log("==================================================");
    console.log(`🎉 抓取完成！`);
    console.log(`✂️  為了讓店家能「垂直」接上馬路，我們打斷並重接了 ${splitEdgesCount} 條道路！`);
    console.log("==================================================");

  } catch (error) {
    console.error("❌ 抓取失敗:", error);
  }
}

fetchAndConnectBuildings();

File: tools/fetch_buildings.py
import requests
import json
import os

# ==========================================
# 1. 環境變數設定
# ==========================================
# 使用你之前設定的校園 Bounding Box
MIN_LON, MIN_LAT = 120.460, 23.550
MAX_LON, MAX_LAT = 120.485, 23.570
OUTPUT_FILE = "ccu_buildings.geojson"

OVERPASS_URL = "https://overpass-api.de/api/interpreter"

def get_building_height(tags):
    """
    根據 OSM 的標籤自動推算建築物高度 (公尺)
    優先級: height -> building:levels -> 預設值
    """
    # 1. 如果有明確標示高度，直接使用
    if 'height' in tags:
        try:
            return float(tags['height'].replace('m', '').strip())
        except ValueError:
            pass
            
    # 2. 如果沒有高度，但有樓層數，每層樓以 3.5 公尺計算
    if 'building:levels' in tags:
        try:
            levels = float(tags['building:levels'])
            return levels * 3.5
        except ValueError:
            pass
            
    # 3. 如果什麼都沒有，預設給 3 層樓的高度 (10.5 公尺)
    return 10.5

def fetch_osm_buildings():
    """
    使用 Overpass QL 查詢範圍內的建築物，並直接取得幾何座標 (geom)
    """
    print("🚀 正在向 Overpass API 請求大樓資料...")
    
    query = f"""
    [out:json][timeout:25];
    (
      way["building"]({MIN_LAT},{MIN_LON},{MAX_LAT},{MAX_LON});
    );
    out body geom;
    """
    
    # 【關鍵修改】加上 Headers 標示身份，避免被伺服器當成惡意爬蟲踢掉
    headers = {
        'User-Agent': 'CCU-Campus-Map-Script/1.0 (dreamgggq@gmail.com)',
        'Accept': '*/*'
    }
    
    # 將 headers 放進請求中
    response = requests.post(OVERPASS_URL, data={'data': query}, headers=headers)
    
    if response.status_code != 200:
        # 加上 response.text，如果再報錯，我們才能看到伺服器具體在抱怨什麼
        raise Exception(f"API 請求失敗: HTTP {response.status_code} - {response.text}")
        
    return response.json()
def convert_to_geojson(osm_data):
    """
    將 OSM 的 JSON 格式轉換為標準的 GeoJSON 格式
    """
    print("🔄 正在轉換為 GeoJSON 格式並計算高度...")
    
    features = []
    
    for element in osm_data.get('elements', []):
        if element['type'] == 'way' and 'geometry' in element:
            # 取得該建築物的標籤 (包含名稱、樓層等)
            tags = element.get('tags', {})
            
            # 計算大樓高度
            height = get_building_height(tags)
            
            # 轉換座標格式：OSM 給 [{'lat': 23.5, 'lon': 120.4}, ...]
            # GeoJSON 需要 [[120.4, 23.5], ...] (注意 GeoJSON 是 [經度, 緯度])
            coordinates = []
            for node in element['geometry']:
                coordinates.append([node['lon'], node['lat']])
                
            # 確保多邊形是閉合的 (起點與終點必須相同)
            if coordinates[0] != coordinates[-1]:
                coordinates.append(coordinates[0])
                
            # 建立 GeoJSON Feature
            feature = {
                "type": "Feature",
                "properties": {
                    "id": element['id'],
                    "name": tags.get('name', '未命名建築'),
                    "building": tags.get('building', 'yes'),
                    "height": height,
                    "levels": tags.get('building:levels', '未知')
                },
                "geometry": {
                    "type": "Polygon",
                    "coordinates": [coordinates] # Polygon 外圍加一層 list
                }
            }
            features.append(feature)
            
    # 組合成完整的 GeoJSON 結構
    geojson = {
        "type": "FeatureCollection",
        "features": features
    }
    
    return geojson

def main():
    try:
        # 1. 下載資料
        osm_data = fetch_osm_buildings()
        
        # 2. 轉換資料
        geojson_data = convert_to_geojson(osm_data)
        
        # 3. 存檔
        with open(OUTPUT_FILE, 'w', encoding='utf-8') as f:
            json.dump(geojson_data, f, ensure_ascii=False, indent=2)
            
        print("=" * 40)
        print("🎉 處理完成！")
        print(f"- 成功抓取了 {len(geojson_data['features'])} 棟大樓的多邊形")
        print(f"- 檔案已儲存為: {OUTPUT_FILE}")
        print("=" * 40)
        
    except Exception as e:
        print(f"❌ 發生錯誤: {e}")

if __name__ == "__main__":
    main()